src/stream/sink/udp_sink.rs

```diff
@@ -1,51 +1,92 @@
+use anyhow::{bail, Context, Result};
 
 use tracing::*;
 
-use gstreamer::prelude::*;
+use gst::prelude::*;
 
-use super::sink::{Sink, SinkInterface};
+use super::sink::SinkInterface;
+use crate::stream::pipeline::pipeline::PIPELINE_TEE_NAME;
 
 #[derive(Debug)]
 pub struct UdpSink {
-    sink_id: String,
-    element: Arc<Mutex<gstreamer::Element>>,
-    sink_pad: gstreamer::Pad,
-    tee_src_pad: Option<gstreamer::Pad>,
+    sink_id: uuid::Uuid,
+    element: gst::Element,
+    sink_pad: gst::Pad,
+    tee_src_pad: Option<gst::Pad>,
 }
 impl SinkInterface for UdpSink {
     #[instrument(level = "debug")]
-    fn get_element(&self) -> std::sync::MutexGuard<'_, gstreamer::Element> {
-        self.element.lock().unwrap()
-    }
+    fn link(
+        &mut self,
+        pipeline: &gst::Pipeline,
+        pipeline_id: &uuid::Uuid,
+        tee_src_pad: gst::Pad,
+    ) -> Result<()> {
+        let sink_id = &self.get_id();
 
-    #[instrument(level = "debug")]
-    fn get_id(&self) -> &String {
-        &self.sink_id
-    }
+        // Add the Sink element to the Pipeline
+        let sink_element = &self.element;
+        if let Err(error) = pipeline.add(sink_element) {
+            bail!("Failed to add UdpSink {sink_id} to Pipeline {pipeline_id}. Reason: {error:#?}");
+        }
 
-    #[instrument(level = "debug")]
-    fn get_sink_pad(&self) -> &gstreamer::Pad {
-        &self.sink_pad
+        // Set Tee's src pad
+        if self.tee_src_pad.is_some() {
+            bail!("Tee's src pad from UdpSink {sink_id} has already been configured");
+        }
+        self.tee_src_pad.replace(tee_src_pad);
+
+        // Link the new Tee's src pad to the Sink's sink pad
+        if let Some(tee_src_pad) = &self.tee_src_pad {
+            if let Err(error) = tee_src_pad.link(&self.sink_pad) {
+                pipeline.remove(sink_element)?;
+                bail!(error)
+            }
+        }
+
+        Ok(())
     }
 
     #[instrument(level = "debug")]
-    fn get_tee_src_pad(&self) -> Option<&gstreamer::Pad> {
-        self.tee_src_pad.as_ref()
+    fn unlink(&self, pipeline: &gst::Pipeline, pipeline_id: &uuid::Uuid) -> Result<()> {
+        let sink_id = self.get_id();
+
+        if let Some(tee_src_pad) = &self.tee_src_pad {
+            if let Err(error) = tee_src_pad.unlink(&self.sink_pad) {
+                bail!("Failed unlinking Sink {sink_id} from Tee's source pad. Reason: {error:?}");
+            }
+
+            let tee = pipeline
+                .by_name(PIPELINE_TEE_NAME)
+                .context(format!("no element named {PIPELINE_TEE_NAME:#?}"))?;
+            if let Err(error) = tee.remove_pad(tee_src_pad) {
+                bail!("Failed removing Tee's source pad. Reason: {error:?}");
+            }
+        }
+
+        let sink_element = &self.element;
+        // if let Err(error) = sink_element.set_state(gst::State::Null) {
+        //     bail!("Failed to set UdpSink state to NULL. Reason: {error:#?}")
+        // }
```

Comment:
```suggestion
```
This was causing a 'bad' failure on gst runtime and doesn't seem to be necessary